processing/src/main/java/org/apache/druid/segment/data/VSizeLongSerde.java

```diff
@@ -413,9 +412,81 @@ public void close() throws IOException
     }
   }
 
+  /**
+   * Unpack bitpacked long values from an underlying contiguous memory block
+   */
   public interface LongDeserializer
   {
+    /**
+     * Unpack long value at the specified row index
+     */
     long get(int index);
+
+    /**
+     * Unpack a contiguous vector of long values at the specified start index of length and adjust them by the supplied
+     * delta base value.
+     */
+    default void getDelta(long[] out, int outPosition, int startIndex, int length, long base)
+    {
+      for (int i = 0; i < length; i++) {
+        out[outPosition + i] = base + get(startIndex + i);
+      }
+    }
+
+    /**
+     * Unpack a non-contiguous vector of long values at the specified indexes and adjust them by the supplied delta base
+     * value.
+     */
+    default int getDelta(long[] out, int outPosition, int[] indexes, int length, int indexOffset, int limit, long base)
```

Comment:
Do you have evidence that the `getDelta` and `getTable` methods are helpful? (vs. the alternative: first calling a regular bulk `get` method, then applying the delta or table adjustment in a loop over the returned arrays)

They complexify the code quite a bit, so we should only include them if they are meaningfully better performance-wise.

Reply:
I haven't measured as you suggest with these exact get methods, (did an earlier test before they were filled out that seemed to show improvement though which is why I went down this path because cutting out the extra vector iteration seemed to be worth the math). i will try to run the numbers this evening to compare.

Reply:
I did the first 8 bits so far without pushdown, the results are 1-2 millis slower.

without pushdown:
```
Benchmark                                                                           (distribution)  (encoding)  (filteredRowCountPercentage)   (rows)
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                    uniform-1    lz4-auto                           1.0  5000000                0.0  avgt    5    20409.851 ±  998.810  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                    uniform-2    lz4-auto                           1.0  5000000                0.0  avgt    5    18935.144 ±  199.459  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                    uniform-4    lz4-auto                           1.0  5000000                0.0  avgt    5    18668.448 ±  538.985  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                    uniform-8    lz4-auto                           1.0  5000000                0.0  avgt    5    19010.981 ±  695.243  us/op
```

compared to with pushdown:
```
Benchmark                                                                           (distribution)  (encoding)  (filteredRowCountPercentage)   (rows)
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                       uniform-1    lz4-auto                           1.0  5000000                0.0  avgt    5     17892.362 ±  102.776  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                       uniform-2    lz4-auto                           1.0  5000000                0.0  avgt    5     17796.103 ±  417.847  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                       uniform-4    lz4-auto                           1.0  5000000                0.0  avgt    5     17879.496 ±  237.066  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                       uniform-8    lz4-auto                           1.0  5000000                0.0  avgt    5     17508.260 ±  560.856  us/op
```

I'll run the rest, but based on the results so far, since only the first 4 deserializers implement `getTable`, I would be in favor of leaving the pushdown in place since it isn't _that_ much extra complexity.

Reply:
without pushdown:

```
Benchmark                                                                        (distribution)  (encoding)  (filteredRowCountPercentage)   (rows)  (zeroProbability)  Mode  Cnt         Score      Error  Units
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-12    lz4-auto                           1.0  5000000                0.0  avgt    5     20574.724 ± 3674.509  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-16    lz4-auto                           1.0  5000000                0.0  avgt    5     20893.434 ± 2894.104  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-20    lz4-auto                           1.0  5000000                0.0  avgt    5     19857.499 ± 1926.944  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-24    lz4-auto                           1.0  5000000                0.0  avgt    5     22194.340 ± 2624.983  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uinform-32    lz4-auto                           1.0  5000000                0.0  avgt    5     18321.336 ±  465.633  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-40    lz4-auto                           1.0  5000000                0.0  avgt    5     23252.329 ±  341.846  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-48    lz4-auto                           1.0  5000000                0.0  avgt    5     25273.632 ± 1414.751  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-56    lz4-auto                           1.0  5000000                0.0  avgt    5     26429.779 ± 2649.011  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                   uniform-64    lz4-auto                           1.0  5000000                0.0  avgt    5     21943.211 ± 2124.446  us/op
```

compare with pushdown:
```
Benchmark                                                                        (distribution)  (encoding)  (filteredRowCountPercentage)   (rows)  (zeroProbability)  Mode  Cnt         Score      Error  Units
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-12    lz4-auto                           1.0  5000000                0.0  avgt    5     18272.440 ±   71.751  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-16    lz4-auto                           1.0  5000000                0.0  avgt    5     19042.292 ±  595.685  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-20    lz4-auto                           1.0  5000000                0.0  avgt    5     18782.746 ±  248.738  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-24    lz4-auto                           1.0  5000000                0.0  avgt    5     19048.354 ±  160.025  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uinform-32    lz4-auto                           1.0  5000000                0.0  avgt    5     17984.778 ±  633.691  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-40    lz4-auto                           1.0  5000000                0.0  avgt    5     22070.007 ±  166.035  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-48    lz4-auto                           1.0  5000000                0.0  avgt    5     22052.517 ± 2168.763  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-56    lz4-auto                           1.0  5000000                0.0  avgt    5     25001.739 ±  259.184  us/op
ColumnarLongsSelectRowsFromGeneratorBenchmark.selectRowsVectorized                      uniform-64    lz4-auto                           1.0  5000000                0.0  avgt    5     20325.369 ±   60.724  us/op
```

Reply:
Thanks for running the tests. It seems worth it to keep it this way.